### DISCUSSION TITLE
What is the "right way" to pass down query data?

### DISCUSSION THREAD

**Discussion by taschetto** [TanStack/query#463] [General] [answered]
A) Props drilling?

B) Creating a custom hook to create a new "instance" of the same query and use it down the component tree?

C) Accessing the query cache directly?

D) Other - which one?

**Comment by tannerlinsley** [comment, marked as answer]
B for subscriptions in components, C for imperative stuff.

**Reply by tannerlinsley** [reply]
Don’t create contexts and don’t pass down query data. Just make a custom hook and utilize the built in context and global cache already in RQ

**Reply by taschetto** [reply]
Wow, thanks for the quick answer. You rock, Tanner. :-)

**Reply by asvny** [reply]
@tannerlinsley @taschetto Wondering if there is any reference for B .. sorry not able to understand what it is.

**Reply by taschetto** [reply]
@asvny You can create a custom hook to reuse your query anywhere:

```javascript
// hooks/useMe.js

import { fetchMe } from 'api/clients/me'
import { useQuery } from 'react-query'

export const useMe = () => useQuery('me', fetchMe)

// api/clients/me.js

import { instance } from 'api'

export const fetchMe = async () => {
  const { data } = await instance.get('/api/auth/me')

  return data
}
```

This way you can use the `useMe` hook at multiple components.

**Reply by asvny** [reply]
Newbie q: Let's say I have data about the current viewer `useCurrentViewer` which is fetched initially on loading in `ViewProvider`.

Should I pass the data in a context for sub-components to consume or should the sub-components use the `useCurrentViewer` hook?
Wouldn't the latter create an extra request for each reference?

BTW .. thanks for your patience.

**Reply by TkDodo** [reply]
Call the hook every time you need the data. It does create a request in the background, but that will make your data more accurate. If you don’t need that, set a `staleTime` for your query.

**Reply by imteekay** [reply]
For anyone looking for the API to get the cached data imperatively, you can use the `getQueryState` function from the `QueryClient`.

```javascript
import { useQueryClient } from 'react-query';

const MyComponent = () => {
  const queryClient = useQueryClient();
  const query = queryClient.getQueryState('THE_KEY');
  // const { data, isFetching, isError, ... } = query;
  ...
};
```

For reference: [docs](https://react-query.tanstack.com/reference/QueryClient#queryclientgetquerystate)

**Reply by julionav** [reply]
I have been trying to do `B`, but this is triggering another query. I mention this in another discussion https://github.com/tannerlinsley/react-query/discussions/2018
